src/sgame/sg_bot_nav.cpp

```diff
@@ -368,6 +368,51 @@ Local Bot Navigation
 ========================
 */
 
+bool BotTraceForFloor( gentity_t *self, uint32_t dir )
+{
+	// this is using a daemon trace to a solid surface
+	// another feasible method might be to do a recast navmesh trace
+	trace_t trace;
+	glm::vec3 dirVec;
+	glm::vec3 viewangles = VEC2GLM( self->client->ps.viewangles );
+	viewangles.x = 0;
+	AngleVectors( viewangles, &dirVec, nullptr, nullptr );
+	// we only handle a subset of the possible values of the bitfield `dir`
+	if ( dir == MOVE_BACKWARD )
+	{
+		dirVec.x = -dirVec.x;
+		dirVec.y = -dirVec.y;
+	}
+	else if ( dir == MOVE_LEFT )
+	{
+		std::swap( dirVec.x, dirVec.y );
+		dirVec.x = -dirVec.x;
+	}
+	else if ( dir == MOVE_RIGHT )
+	{
+		std::swap( dirVec.x, dirVec.y );
+		dirVec.y = -dirVec.y;
+	}
+	else
+	{
+		return false;
+	}
+	glm::vec3 playerMins, playerMaxs;
+	class_t pClass = static_cast<class_t>( self->client->ps.stats[STAT_CLASS] );
+	BG_BoundingBox( pClass, &playerMins, &playerMaxs, nullptr, nullptr, nullptr );
+	float radius = playerMaxs.x - playerMins.x;
+	float height = playerMaxs.z - playerMins.z;
+	// use dirVec * ( 2.f * radius ) as direction offset
+	// in theory, sqrtf(2.f) instead of 2.f would be enough
+	// the slight increase is to safeguard against unexpected map geometry
+	// a bot not dancing is less problematic than a bot falling down
+	glm::vec3 start = VEC2GLM( self->s.origin ) + dirVec * ( 2.f * radius );
+	glm::vec3 end = start;
+	end.z -= height;
```

Comment:
By doing a trace of length of `playerMaxs.z` - `playerMins.z` that starts at a height of `-playerMins.z` above the player's feet, you end up tracing a distance of `playerMaxs.z` below the feet. It's not good to write code that depends on `playerMaxs.z` because the height of the origin within the bbox is arbitrary. The max z could be anywhere from 0 to the entire height of the bbox without affecting collisions.

To me it would make more sense to trace a constant distance below the player's feet.

Reply:
Do I get the player's feet height by doing

```
glm::vec3 origin = VEC2GLM( self->s.origin );
origin.z += playerMins.z;
```

Reply:
That's right

Reply:
> To me it would make more sense to trace a constant distance below the player's feet.

That would miss any upward slopes. Better trace from the player's top.

Reply:
True, if you are doing a point trace you would need to start higher.